diesel/src/query_source/aliasing/alias.rs

```diff
@@ -0,0 +1,171 @@
+use super::{AliasSource, AliasedField, FieldAliasMapper};
+
+use crate::associations::HasTable;
+use crate::backend::Backend;
+use crate::expression::{Expression, SelectableExpression, ValidGrouping};
+use crate::query_builder::nodes::StaticQueryFragment;
+use crate::query_builder::{AsQuery, AstPass, FromClause, QueryFragment, QueryId, SelectStatement};
+use crate::query_source::{AppearsInFromClause, Column, Never, QuerySource, Table, TableNotEqual};
+use crate::result::QueryResult;
+
+#[derive(Debug, Clone, Copy, Default)]
+/// Represents an alias within diesel's query builder
+///
+/// See [alias!] for more details.
+pub struct Alias<S> {
+    source: S,
+}
+
+impl<S: AliasSource> Alias<S> {
+    /// Maps a single field of the source table in this alias
+    pub fn field<F>(&self, field: F) -> AliasedField<S, F>
+    where
+        S: Clone,
+        F: Column<Table = S::Table>,
+    {
+        AliasedField {
+            _alias_source: self.source.clone(),
+            _field: field,
+        }
+    }
+    /// Maps multiple fields of the source table in this alias
+    /// (takes in tuples and some expressions)
+    pub fn fields<Fields>(&self, fields: Fields) -> <Fields as FieldAliasMapper<S>>::Out
+    where
+        Fields: FieldAliasMapper<S>,
+    {
+        fields.map(self)
+    }
+}
+
+impl<S> Alias<S> {
+    #[doc(hidden)]
+    /// May be used to create an alias. Used by the [`alias!`] macro.
+    pub const fn new(source: S) -> Self {
+        Self { source }
+    }
+}
+
+impl<S> QueryId for Alias<S>
+where
+    Self: 'static,
+    S: AliasSource,
+    S::Table: Table,
+{
+    type QueryId = Self;
+    const HAS_STATIC_QUERY_ID: bool = true;
```

Comment:
That should be depended on `AliasSource::Table`. If that has a static query id, that's here the case as well. If not than the query id of the whole alias is not stable.

Reply:
Fine for me to add that (I'll do it), but I thought that was already implied by the fact it was a `Table` (hence supposedly generated by the `table!` macro) -> why do we prefer writing it like this?

Reply:
`diesel_dynamic_schema::Table` implements `Table` as well, but that does not have a static query id for obvious reasons.

https://github.com/diesel-rs/diesel/blob/master/diesel_dynamic_schema/src/table.rs#L75